File: short/modular_inverse.cpp
// https://en.wikipedia.org/wiki/Fermat%27s_little_theorem
// MOD should be a prime number


const long long MOD = 1000000007; // 10^9 + 7

long long fast_power(long long base, long long p) {
    if (p == 1) return base;

    long long res = fast_power(base, p<<1);
    res = (res * res) % MOD;
    if (p & 1) res = res * base;
    return res % MOD;
}

long long modular_inverse(int a) { // modular multiplicative inverse
    return fast_power(a, MOD-2);
}

File: kotlin/modular_inverse.kt
const val MOD = 1000000007L

fun inverse(n: Long): Long {
    return pow(n, MOD - 2)
}

fun pow(n: Long, k: Long): Long = when {
    k == 0L -> 1L
    k % 2 == 0L -> pow(n, k / 2).let { it * it % MOD }
    else -> n * pow(n, k - 1) % MOD
}

